LibOS/shim/src/shim_malloc.c

```diff
@@ -65,197 +49,51 @@ static LOCKTYPE slab_mgr_lock;
-        if (!addr) {
-            bkeep_munmap(addr, alloc_size, flags);
-            unlock(shim_heap_lock);
-            return NULL;
-        }
-    } else {
-
-        struct shim_heap * heap = __alloc_enough_heap(alloc_size);
-
-        if (!heap) {
-            unlock(shim_heap_lock);
-            return NULL;
-        }
+    addr = (void *) DkVirtualMemoryAlloc(addr, alloc_size, 0,
+                                         PAL_PROT_WRITE|PAL_PROT_READ);
 
-        addr = heap->current;
-        heap->current += alloc_size;
+    if (!addr) {
+        bkeep_munmap(addr, alloc_size, flags);
+        return NULL;
     }
 
-    unlock(shim_heap_lock);
-
     return addr;
 }
 
 void __system_free (void * addr, size_t size)
 {
-    int in_reserved_area = 0;
-    DkVirtualMemoryFree(addr, ALIGN_UP(size));
-    int flags = VMA_INTERNAL;
-    for (int i = 0 ; i < MAX_SHIM_HEAP_AREAS ; i++)
-        if (shim_heap_areas[i].start) {
-            /* Here we assume that any allocation from the 
-             * shim_heap_area is a strict inclusion.  Allocations
-             * cannot partially overlap.
-             */
-            if (addr >= shim_heap_areas[i].start
-                && addr <= shim_heap_areas[i].end)
-                in_reserved_area = 1;
-        }
-
-    if (!in_reserved_area)
-        bkeep_munmap(addr, ALIGN_UP(size), flags);
-}
-
-int init_heap (void)
-{
-    create_lock(shim_heap_lock);
-
-    void * start = (void *) DkVirtualMemoryAlloc(NULL, INIT_SHIM_HEAP, 0,
-                                    PAL_PROT_WRITE|PAL_PROT_READ);
-    if (!start)
-        return -ENOMEM;
-
-    debug("allocate internal heap at %p - %p\n", start,
-          start + INIT_SHIM_HEAP);
-
-    shim_heap_areas[0].start = shim_heap_areas[0].current = start;
-    shim_heap_areas[0].end = start + INIT_SHIM_HEAP;
-
-    return 0;
-}
+    if (bkeep_munmap(addr, ALIGN_UP(size), VMA_INTERNAL) < 0)
+        return;
 
-int bkeep_shim_heap (void)
-{
-    lock(shim_heap_lock);
-    
-    for (int i = 0 ; i < MAX_SHIM_HEAP_AREAS ; i++)
-        if (shim_heap_areas[i].start) {
-            /* Add a VMA for the active region */
-            bkeep_mmap(shim_heap_areas[i].start,
-                       shim_heap_areas[i].current - shim_heap_areas[i].start,
-                       PROT_READ|PROT_WRITE,
-                       MAP_PRIVATE|MAP_ANONYMOUS|VMA_INTERNAL, NULL, 0, "heap");
-            /* Go ahead and free the reserved region */
-            if (shim_heap_areas[i].current < shim_heap_areas[i].end) {
-                DkVirtualMemoryFree(shim_heap_areas[i].current,
-                                    ALIGN_UP(((long unsigned int) shim_heap_areas[i].end) - ((long unsigned int) shim_heap_areas[i].current)));
-                shim_heap_areas[i].end = shim_heap_areas[i].current;
-            }
-        }
-    vmas_initialized = 1;
-    
-    unlock(shim_heap_lock);
-    return 0;
+    DkVirtualMemoryFree(addr, ALIGN_UP(size));
```

Comment:
Shouldn't we call this function *before* publishing that the area is free?

Reply:
This sounds right to me.

Reply:
Done.